packages/slate-react/src/components/string.tsx

```diff
@@ -59,21 +59,30 @@ const TextString = (props: { text: string; isTrailing?: boolean }) => {
   const { text, isTrailing = false } = props
 
   const ref = useRef<HTMLSpanElement>(null)
-  const forceUpdateCount = useRef(0)
 
-  if (ref.current && ref.current.textContent !== text) {
-    forceUpdateCount.current += 1
-  }
+  // This is the actual text rendering boundary where we interface with the DOM
+  // The text is not rendered as part of the virtual DOM, as since we handle basic character insertions natively,
+  // updating the DOM is not a one way dataflow anymore. What we need here is not reconciliation and diffing
+  // with previous version of the virtual DOM, but rather diffing with the actual DOM element, and replace the DOM <span> content
+  // exactly if and only if its current content does not match our current virtual DOM.
+  // Otherwise the DOM TextNode would always be replaced by React as the user types, which interferes with native text features,
+  // eg makes native spellcheck opt out from checking the text node.
 
-  // This component may have skipped rendering due to native operations being
-  // applied. If an undo is performed React will see the old and new shadow DOM
-  // match and not apply an update. Forces each render to actually reconcile.
-  return (
-    <span data-slate-string ref={ref} key={forceUpdateCount.current}>
-      {text}
-      {isTrailing ? '\n' : null}
-    </span>
-  )
+  // useLayoutEffect: updating our span before browser paint
+  useLayoutEffect(() => {
+    // null coalescing text to make sure we're not outputing "null" as a string in the extreme case it is nullish at runtime
+    const textWithTrailing = (text ?? '') + (isTrailing ? '\n' : '')
```

Comment:
For your consideration, maybe we do interpolation here instead?

```suggestion
    const textWithTrailing = `${text ?? ''}${isTrailing ? '\n' : ''}`
```

I don't think it's better or worse, just more ES6-ish. Not a factor for merging the PR though.